### PR TITLE
RFC5280 strict time string check

### LINKED ISSUE #3444: Generate valid time values in ca program
I just found 2 bugs reports that point out that the openssl ca app just puts -startdate and -enddate into the certificate as it was specified on the command line even when it was invalid. We should either reject them or normalize them.

Bugs:
https://gitlab.com/gnutls/gnutls/issues/196
https://bugs.debian.org/862335

### PR DESCRIPTION
**Pull request by InfoHunter**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

Fixes #3444 

Thoughts: introduce a new API called "ASN1_TIME_set_string_strict" which is used to check if the time string (both UTC and Generalized) matches what RFC 5280 requires.

To make it compatible with those legacy certificates generated by previous versions of OpenSSL command line tool, all use the original "loose" rule to check the time string except the 'ca' application which is enforced to use the strict version to produce RFC-5280-fulfilled certificates only. Thus, `openssl` command can still read and use RFC-5280-time-conficting certificates but can't generate any of them.

Docs and tests will be updated as this fix is generally accepted for the concept/idea.

Please ignore the first unrelated 3 outdated commits : )

### CODE REVIEW

**Comment by kroeckx** [comment]
I guess I would prefer a function in the X509 name space and not in the generic ASN1. It's for X509 that we want to be strict. But then I'm not sure it's important.

I would also like to see tests to see that we reject invalid strings now, and that it properly ends up as UTC / Generalized.

**Comment by InfoHunter** [comment]
Test cases updated.

**Comment by InfoHunter** [comment]
Hi @kroeckx 

I changed ASN1_TIME_set_string_strict to ASN1_TIME_set_string_X509, it's still in the ASN1 name space but the function name is more specific to X509. Since all X509 time related stuff is handled by ASN1 time functions now in OpenSSL, so it might be 'OK' to consider the strict X509 format as a a 'subset' feature offered by the underlaying ASN1. But it still looks good if we add an X509_time_xxx function in x509_vfy.c to wrap the ASN1 layer, so what do you think?

Documentation and test cases are all updated. Please let me know your considerations.

File: test/x509_time_test.c [hunk 1]
```diff
@@ -25,6 +25,46 @@ typedef struct {
...
 } TESTDATA;
 
+/*
+ * Actually, the "loose" mode has been tested in
+ * those time-compare-cases, so we may not test it again.
+ */
+static TESTDATA x509_strict_tests[] = {
+    {
+        "20170217180154Z", V_ASN1_GENERALIZEDTIME,
+        0, 1,
```
Comment (kroeckx) [inline test/x509_time_test.c]: I think you should probably either add something to the TESTDATA structure or use a different one. The current expected is the result of a compare, so you have a cmp_time and when comparing the parsed and the expected value you expect to get something. You don't use the cmp_time, and use expected for something else.

You also don't seem to be using the type you set in the test data, and adjust the type based on the string you give, which doesn't make sense. I think you should check both that parsing is successful and that it returned the correct type. The correct type for all of your strings should be V_ASN1_UTCTIME, because they're all before 2050. You should also create strings that are after 2050.

Comment (kroeckx) [inline test/x509_time_test.c]: It might also be useful to do a check like giving "20170217180154Z" and checking that it turned it into "170217180154Z" (and is a V_ASN1_UTCTIME)

Comment (InfoHunter) [inline test/x509_time_test.c]: Hmm, I know what's going wrong: previously I only considered 'checking' the time string format but what is described in the 4.1.2.5. section of RFC 5280 is not implemented in my patch. So the test cases cover only the 'check' part, but no 're-format' the time string.

I need a little more time to add the missing functionality, thanks man!

**Comment by InfoHunter** [comment]
Hi @kroeckx , I have update the code to have "2050-Generalized-to-UTC" feature implemented, as well as test cases and documentation. Please review them, thanks.

File: include/openssl/asn1.h [hunk 2]
```diff
@@ -141,6 +141,8 @@ DEFINE_STACK_OF(X509_ALGOR)
 # define ASN1_STRING_FLAG_MSTRING 0x040
 /* String is embedded and only content should be freed */
 # define ASN1_STRING_FLAG_EMBED 0x080
+/* String should be parsed in RFC 5280's time format */
+# define ASN1_STRING_FLAG_X509_TIME 0x100
```
Comment (kroeckx) [inline include/openssl/asn1.h]: I've been wondering if we shouldn't just set this flag somehow instead of creating a new API just to set this flag. I don't actually know enough about the whole API to know how to extend it best, so I'm open for other ideas.

File: test/x509_time_test.c [hunk 3]
```diff
@@ -187,8 +280,53 @@ static int test_x509_cmp_time_current()
...
+    result = ASN1_TIME_set_string_X509(t, x509_format_tests[idx].data);
+    /* time string parsing result is always checked against what's expected */
+    if (!TEST_int_eq(result, x509_format_tests[idx].expected)) {
+        TEST_info("test_x509_time(%d) failed: expected %d, got %d\n",
+                idx, x509_format_tests[idx].expected, result);
+        return 0;
+    }
+
+    /* if t is not NULL but expected_type is ignored(-1), it is an 'OK' case */
+    if (t && x509_format_tests[idx].expected_type != -1) {
```
Comment (kroeckx) [inline test/x509_time_test.c]: use t != NULL

File: test/x509_time_test.c [hunk 4]
```diff
@@ -187,8 +280,53 @@ static int test_x509_cmp_time_current()
...
+    int result;
+
+    if (x509_format_tests[idx].set_string) {
+        /* set-string mode */
+        t = ASN1_TIME_new();
+        if (t == NULL) {
+            TEST_info("test_x509_time(%d) failed: internal error\n", idx);
+            return 0;
+        }
+    }
```
Comment (kroeckx) [inline test/x509_time_test.c]: I don't really understand the point of this, why not always give it an ASN1_TIME?

File: test/x509_time_test.c (hunk 3 above)
Comment (InfoHunter) [inline test/x509_time_test.c]: OK

File: test/x509_time_test.c (hunk 4 above)
Comment (InfoHunter) [inline test/x509_time_test.c]: Because we need to test if this function can accept a NULL input for the first parameter and if behaves properly. The doc under doc/man3/ASN1_TIME_set.pod didn't describe this usage for the original ASN1_TIME_set_string(), but it's indeed used that way in apps/ca.c and might be used already by other applications. So I think we need to cover this scenario in case of the future-API-internal-change, we would probably get noticed if input of NULL ASN1_TIME  still works or not.

File: include/openssl/asn1.h (hunk 2 above)
Comment (InfoHunter) [inline include/openssl/asn1.h]: IMHO, since we need to utilize the ASN1_TIME to struct tm parsing logic, it might be one of the best ways to reduce duplicated code if we use a flag and treat it in current  ASN1_TIME to struct tm parsing functions.

**Comment by InfoHunter** [comment]
Is there a way to re-launch the travis-ci manually?

It seems it failed because the server didn't start:
```
Client command: echo test | ../util/shlib_wrap.sh ../apps/openssl s_client -engine ossltest -connect localhost:4453 -ct
engine "ossltest" set.
Connection opened
Failed to start up server (localhost,4443): Connection refused
Server command: ../util/shlib_wrap.sh ../apps/openssl s_server -no_comp -rev -engine ossltest -accept 4443 -cert ../apps/server.pem -cert2 ../apps/server.pem -naccept 1 -cipher AES128-SHA:TLS13-AES-128-GCM-SHA256 -status_file ../test/recipes/ocsp-response.der -serverinfo ../test/serverinfo2.pem
```

Per 9fa299f

**Comment by levitte** [comment]
Okie, job restarted

**Comment by InfoHunter** [comment]
@levitte , is this normal? I mean a server can't start

**Comment by mattcaswell** [comment]
> @levitte , is this normal? I mean a server can't start

Looking at the logs it seems to me that the server *did* actually start eventually. If you look at the following test the server fails to start there too because the address is already in use. The problem seems to be we gave up trying to connect too quickly. Looking at the code we try 3 times with a 0.1 sec pause between each attempt. Perhaps that is too aggressive.

**Comment by InfoHunter** [comment]
Yep, I noticed those `bind` errors. I agree with you that an 0.1s internal is too aggressive. Should it be adjust to avoid this? Or maybe the number of tries should be increased.

**Comment by mattcaswell** [comment]
> Yep, I noticed those bind errors. I agree with you that an 0.1s internal is too aggressive. Should it be adjust to avoid this? Or maybe the number of tries should be increased.

I upped the retries: #3587

File: include/openssl/asn1.h (hunk 2 above)
Comment (levitte) [inline include/openssl/asn1.h]: There is no precedent, all other flags are set internally, so I think we're at liberty to define how to set this particular flag whatever way we wish.  The way it's implemented here seems sane to me.

File: doc/man3/ASN1_TIME_set.pod [hunk 5]
```diff
@@ -11,6 +12,7 @@ ASN1_TIME_print, ASN1_TIME_diff - ASN.1 Time functions
  ASN1_TIME *ASN1_TIME_adj(ASN1_TIME *s, time_t t,
                           int offset_day, long offset_sec);
  int ASN1_TIME_set_string(ASN1_TIME *s, const char *str);
+ int ASN1_TIME_set_string_X509(ASN1_TIME *s, const char *str);
```
Comment (levitte) [inline doc/man3/ASN1_TIME_set.pod]: Please add a history section mentioning this was added for OpenSSL 1.1.1.  There are examples on how to do this in other .pod files

File: doc/man3/ASN1_TIME_set.pod [hunk 6]
```diff
@@ -32,7 +34,16 @@ and returned.
...
+format (for example YYMMDDHHMMSSZ or YYYYMMDDHHMMSSZ). If B<s> is NULL
+this function performs a format check on B<str> only.
+
+ASN1_TIME_set_string_X509() is more 'strict' than ASN1_TIME_set_string()
+while checking the string B<str>'s format against RFC 5280's rules. For
+instance, SS and Z must be present in a UTCTime format and fractional
+seconds are not allowed in a GeneralizedTime format, and in both format,
+time zones provided by + or - are not allowed. Besides, if the year presented
+by B<str> is between 1950 (included) and 2049, the ASN1_TIME structure B<s>
+is set in UTCTime, otherwise it will be set to GeneralizedTime.
```
Comment (kroeckx) [inline doc/man3/ASN1_TIME_set.pod]: I think we don't need to document in detail what it all checks, just that it's restricts it to what RFC5280 allows. So maybe it should say that it only allows YYMMDDHHMMSSZ and YYYYMMDDHHMMSSZ, while the other functions allows others.

**Comment by kroeckx** [comment]
So it seems there is a conflict that you will need to resolve.

File: crypto/asn1/a_time.c [hunk 7]
```diff
@@ -130,6 +129,62 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
...
+     * 2. UTC: YYMMDDHHMMSSZ, if YY >= 50 (19YY) --> UTC: YYMMDDHHMMSSZ
+     * 3. G'd: YYYYMMDDHHMMSSZ, if YYYY >= 2050 --> G'd: YYYYMMDDHHMMSSZ
+     * 4. G'd: YYYYMMDDHHMMSSZ, if YYYY < 2050 --> UTC: YYMMDDHHMMSSZ
+     *
+     * Only strings of the 4th rule should be reformatted, but since a
+     * UTC can only present [1950, 2050), so if the given time string
+     * is less than 1950 (e.g. 19230419000000Z), we do nothing...
+     */
+
+    if (s && t.type == V_ASN1_GENERALIZEDTIME) {
```
Comment (kroeckx) [inline crypto/asn1/a_time.c]: s != NULL

File: crypto/asn1/a_time.c [hunk 8]
```diff
@@ -130,6 +129,62 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
...
+            t.length -= 2;
+            t.data = OPENSSL_zalloc(t.length + 1);
+            if (t.data == NULL)
+                goto out;
+            memcpy(t.data, str + 2, t.length);
+            t.type = V_ASN1_UTCTIME;
+        }
+    } else {
+        /* t.type == V_ASN1_UTCTIME or s == NULL, everything's fine */
+    }
```
Comment (kroeckx) [inline crypto/asn1/a_time.c]: This else is useless

File: crypto/asn1/a_time.c [hunk 9]
```diff
@@ -130,6 +129,62 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
...
+
+    if (!s || ASN1_STRING_copy((ASN1_STRING *)s, (ASN1_STRING *)&t))
```
Comment (kroeckx) [inline crypto/asn1/a_time.c]: s == NULL

File: crypto/asn1/a_time.c [hunk 10]
```diff
@@ -130,6 +129,62 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
...
+     * UTC can only present [1950, 2050), so if the given time string
+     * is less than 1950 (e.g. 19230419000000Z), we do nothing...
+     */
+
+    if (s && t.type == V_ASN1_GENERALIZEDTIME) {
+        if (!asn1_generalizedtime_to_tm(&tm, &t))
+            goto out;
+        if (tm.tm_year >= 50 && tm.tm_year < 150) {
+            t.length -= 2;
+            t.data = OPENSSL_zalloc(t.length + 1);
```
Comment (kroeckx) [inline crypto/asn1/a_time.c]: I'm wondering if this introduces some memory leak. Should we free str in this case?

Comment (kroeckx) [inline crypto/asn1/a_time.c]: We should probably at least document the transfer of ownership of the pointer that is passed.

File: crypto/asn1/a_gentm.c [hunk 11]
```diff
@@ -22,7 +22,7 @@ int asn1_generalizedtime_to_tm(struct tm *tm, const ASN1_GENERALIZEDTIME *d)
     static const int min[9] = { 0, 0, 1, 1, 0, 0, 0, 0, 0 };
     static const int max[9] = { 99, 99, 12, 31, 23, 59, 59, 12, 59 };
```
Comment (eabalea) [inline crypto/asn1/a_gentm.c]: I had never tried this before, but 23:59:60 is a valid time, so max[5] should be 60.

File: doc/man3/ASN1_TIME_set.pod (hunk 5 above)
Comment (InfoHunter) [inline doc/man3/ASN1_TIME_set.pod]: Roger.

File: doc/man3/ASN1_TIME_set.pod (hunk 6 above)
Comment (InfoHunter) [inline doc/man3/ASN1_TIME_set.pod]: Okay, I'll make it briefer.

File: crypto/asn1/a_time.c (hunk 7 above)
Comment (InfoHunter) [inline crypto/asn1/a_time.c]: Sure. So the 'explicit way' of checking a NULL pointer is preferred all through the whole OpenSSL?

File: crypto/asn1/a_time.c (hunk 8 above)
Comment (InfoHunter) [inline crypto/asn1/a_time.c]: Yep, deleted

File: crypto/asn1/a_time.c (hunk 10 above)
Comment (InfoHunter) [inline crypto/asn1/a_time.c]: The new 't.data' is freed later after ASN1_STRING_copy is done, and we don't care about the old 't.data' since it's passed here from outer function. I will put a comment on this.

File: crypto/asn1/a_gentm.c (hunk 11 above)
Comment (InfoHunter) [inline crypto/asn1/a_gentm.c]: Hmm, I dug a little on 'leap second' and I don't think '23:59:60' is valid with all date. In fact, there are only specific dates in which 'leap second' is valid, such as 2016.12.31, 2015.6.30... So I suspect that we may need extra efforts to check if 'date' is also valid instead of just increasing the max-second, otherwise we may generate a certificate with the validity date '2017.6.1 23:59:60' which is definitely an invalid date/time.

**Comment by InfoHunter** [comment]
@kroeckx @levitte 

Updated according to review.

**Comment by richsalz** [comment]
That’s a good point.  IIRC there can be two leap-seconds, so 61.  Sigh.

**Comment by InfoHunter** [comment]
It's interesting. I find that a 'leap second' was always applied when the gap between UTC and the mean solar time was approaching to 0.9s. So theoretically there should not be 2 leap seconds applied at one time. But if someday the earth rotates itself suddenly very slow and there's no time for the guys at relevant 'time institute' to trigger a lead second, it will happen (and possibly world ends... : )

Seriously,  back to implementation, I think it's impossible to judge accurately a in-future-leap-second-time is valid or invalid. But we can check if the date is on June.30 or Dec.31. I've no idea whether this makes sense or not.

File: crypto/asn1/a_gentm.c (hunk 11 above)
Comment (eabalea) [inline crypto/asn1/a_gentm.c]: The leap second can be inserted at the end of any month (preferably in June and December). I was trying to spot the fact that 20161231235960Z is not considered valid with the current code, while 20160231120000Z is considered valid. And there can be only one leap second inserted (or removed). "2017.6.1 23:59:60" is a valid date (wrt ISO8601), it just does not exist, that's different.

**Comment by kroeckx** [comment]
Since we can't predict the future, I suggest we always reject leap
seconds. We could in theory support it for past dates, but I don't
see the point.

**Comment by richsalz** [comment]
I agree about rejecting leap seconds.  Just note it in the documentation.

File: crypto/asn1/a_gentm.c (hunk 11 above)
Comment (InfoHunter) [inline crypto/asn1/a_gentm.c]: I get your point. It makes sense if the current code can mark date string like '20160231120000Z' as invalid and it might be fixed. But you would probably agree with me that we should not accept '20161231235960Z' just because the current code accepts '20160231120000Z'-like stuff. So I would like to leave it in this PR and might solve this later in other thread.

**Comment by InfoHunter** [comment]
Documentation has been updated to document leap second rejection. Ping @kroeckx and @richsalz .
"History" section has been added to doc to fulfill your change-request, ping @levitte

**Comment by levitte** [comment]
I'm satisfied with the docs, refering the rest to @kroeckx

**Comment by InfoHunter** [comment]
ping @kroeckx , any further considerations on this PR?

**Comment by InfoHunter** [comment]
Solve conflicts with master branch.

I also noticed that the PR #3378 has changed some same files, so I will then solve that conflicts after PR #3378 is merged into master.

**Comment by richsalz** [comment]
ping @kroeckx

**Comment by InfoHunter** [comment]
Conflicts are solved, now it is compatible with the newly added 'ASN1_TIME_to_tm' function : )

**Comment by InfoHunter** [comment]
Hmmm, it seems some tests in 70-test_tls13kexmodes.t didn't run:

```
Forwarded packet length = 28

   0 items in the session cache
   0 client connects (SSL_connect())
   0 client renegotiates (SSL_connect())
   0 client connects that finished
   1 server accepts (SSL_accept())
   0 server renegotiates (SSL_accept())
   1 server accepts that finished
   1 session cache hits
   0 session cache misses
   0 session cache timeouts
   0 callback cache hits
   0 cache full overflows (128 allowed)
Failed 2/11 subtests 
```

**Comment by richsalz** [comment]
It looks like travis hiccup’d.  Force/push a commit (edit your message, for example)

**Comment by InfoHunter** [comment]
Hmmm, hiccup seems cured.

**Comment by InfoHunter** [comment]
Squash all previous commits into one.

File: crypto/asn1/a_gentm.c [hunk 12]
```diff
@@ -88,6 +104,9 @@ int asn1_generalizedtime_to_tm(struct tm *tm, const ASN1_GENERALIZEDTIME *d)
      * digits.
      */
     if (a[o] == '.') {
+        if (strict)
+            /* RFC 5280 forbids fractional seconds */
+            goto err;
         if (++o > l)
             goto err;
```
Comment (kroeckx) [inline crypto/asn1/a_gentm.c]: This check doesn't seem to make sense to me. After doing the `++o`, o is equal to the number of characters processed. if `o == l` it means we can't read any more characters. If `o > l` it means we already read 1 character too much. As far as I can see we always NUL terminate a string, so this is no problem, but then the above check `a[o] == '.'` would have failed. Am I missing something?

File: crypto/asn1/a_gentm.c [hunk 13]
```diff
@@ -100,7 +119,7 @@ int asn1_generalizedtime_to_tm(struct tm *tm, const ASN1_GENERALIZEDTIME *d)
 
     if (a[o] == 'Z')
         o++;
-    else if ((a[o] == '+') || (a[o] == '-')) {
+    else if (!strict && ((a[o] == '+') || (a[o] == '-'))) {
```
Comment (kroeckx) [inline crypto/asn1/a_gentm.c]: There doesn't seem to be a check that the Z is actually present. I think it currently accepts any character as the last character.

Comment (kroeckx) [inline crypto/asn1/a_gentm.c]: Oh, I guess you end up at the /* Missing time zone information. */ in that case, which is fine.

Comment (kroeckx) [inline crypto/asn1/a_gentm.c]: This should probably check that `o < l` before looking at a[o].

File: crypto/asn1/a_time.c [hunk 14]
```diff
@@ -130,6 +129,65 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
...
+            t.length -= 2;
+            /*
+             * it's OK to let original t.data go since that's assigned
+             * to a piece of memory allocated outside of this function.
+             * new t.data would be freed after ASN1_STRING_copy is done.
+             */
+            t.data = OPENSSL_zalloc(t.length + 1);
+            if (t.data == NULL)
+                goto out;
+            memcpy(t.data, str + 2, t.length);
```
Comment (kroeckx) [inline crypto/asn1/a_time.c]: You should probably set t.data[t.length] to '\0'

File: util/libcrypto.num [hunk 15]
```diff
@@ -4298,3 +4298,4 @@ UI_dup_user_data                        4240	1_1_1	EXIST::FUNCTION:UI
 UI_method_get_data_destructor           4241	1_1_1	EXIST::FUNCTION:UI
 ERR_load_strings_const                  4242	1_1_1	EXIST::FUNCTION:
 ASN1_TIME_to_tm                         4243	1_1_1	EXIST::FUNCTION:
+ASN1_TIME_set_string_X509               4244 	1_1_1	EXIST::FUNCTION:
```
Comment (kroeckx) [inline util/libcrypto.num]: Do we want to backport this to 1.1.0? It should get version 1_1_0g in that case.

**Comment by kroeckx** [comment]
I guess I would prefer that the asn1_generalizedtime_to_tm() and asn1_utctime_to_tm() would be one function that just did the right thing depending on the type instead of 2 functions that are almost identical but slightly different. I would also like to see the function documented, and it should then probably document that it always returns the time in UTC and that it drops fractional seconds.

File: crypto/asn1/a_gentm.c (hunk 12 above)
Comment (InfoHunter) [inline crypto/asn1/a_gentm.c]: I think you're right, the check will never be true. Currently it's a 'add-then-check' mode while iterating the input time string, the check is done by comparing the value in a[o] so it works only on null-terminated strings. For this check, probably just change `++o > l` to `++o == l` is fine enough, to avoid unnecessary subsequent checks. But I would like to make this logic more clear to change it to a 'non-null-terminated-depending' style, thus we don't need to be bothered by if the `t->data` has a '\0' or not any more. And I consider this as a more careful and more secure way of dealing with strings.

The same situations exist as well as in UTC parsing function.

File: crypto/asn1/a_gentm.c (hunk 13 above)
Comment (InfoHunter) [inline crypto/asn1/a_gentm.c]: Not necessary, since the o at most can point to the '\0' character.

File: crypto/asn1/a_time.c (hunk 14 above)
Comment (InfoHunter) [inline crypto/asn1/a_time.c]: It's already set by OPNESSL_zalloc.

File: util/libcrypto.num (hunk 15 above)
Comment (InfoHunter) [inline util/libcrypto.num]: From my understandings, it seems that 1.1.0 does not accept new features. Even if we backport it to 1.1.0, I don't know if the patch can merge into that branch without conflicts.

Comment (levitte) [inline util/libcrypto.num]: We have added some new functions in 1.1.0, but those were "things we forgot before the initial release and that really *should* have been there" or "we can't fix this bug without this function".

I'm not sure this PR is of either kind.

**Comment by InfoHunter** [comment]
I agree. But I consider this as a 'refactor' to those functions to reduce duplicated code. Since it needs a bit more changes to the code, and it's not really related to the problem we aim to solve ( Issue #3444 ), I'd rather do this by submit an separate patch later (probably along with other 'refactor' changes that I also have found in other code). So does this strategy make sense to you? @kroeckx

File: crypto/asn1/a_time.c (hunk 14 above)
Comment (kroeckx) [inline crypto/asn1/a_time.c]: I didn't notice it was zalloc ...

File: util/libcrypto.num (hunk 15 above)
Comment (kroeckx) [inline util/libcrypto.num]: As with many things you can argue that this is a bug or not. I'm fine with 1.1.1 only.

**Comment by kroeckx** [comment]
An other PR works for me. An other thing I would like to see if to check that the number of days in the month is actually valid, and I suggest you do that after turning it into 1 function.

**Comment by InfoHunter** [comment]
I pushed a new commit to change the original logic of parsing a time string into 'non-null-terminated-depending'. I didn't squash this commit into the previous one is because I don't think they are 'much' related. It's somehow an 'optimization' of the string parsing. review request, @kroeckx @richsalz @levitte

**Comment by kroeckx** [comment]
@richsalz: Are you still happy with the current version?

**Comment by richsalz** [comment]
I'm still okay; let me know if you want me to merge.

**Comment by kroeckx** [comment]
Feel free to merge

**Comment by richsalz** [comment]
merged and squashed.  thanks, paul!
